File: medium/78_subsets.cpp
class Solution {
public:
    // Time: O(2^n)
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> solution;
        vector<vector<int>> result;
        findSubsets(result, solution, nums, 0);
        return result;
    }
    
    void findSubsets(vector<vector<int>> &res, vector<int> sol, const vector<int> &nums, int idx) {
        res.push_back(sol);
        for (int i = idx; i < nums.size(); ++i) {
            sol.push_back(nums[i]);
            findSubsets(res, sol, nums, i + 1); // attention! not idx + 1!
            sol.pop_back();
        }
    }
    
    // DFS recursive
    /*
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> solution;
        vector<vector<int>> result;
        findSubsets(result, solution, nums, 0);
        return result;
    }
    
    void findSubsets(vector<vector<int>> &res, vector<int> &sol, const vector<int> &nums, int idx) {
        if(idx == nums.size()) {
            res.push_back(sol);
            return;
        }
        
        findSubsets(res, sol, nums, idx + 1);
        sol.push_back(nums[idx]);
        findSubsets(res, sol, nums, idx + 1);
        sol.pop_back();
    }
    */
};

File: medium/78_subsets.java
class Solution {
    // Backtrace to list all the subsets
    // Time: O(2^n) 每个数字都有在不在subset内的选择
    // Space: O(n^2)
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> results = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        backtrace(results, subset, nums, 0);
        return results;
    }
    
    private void backtrace(List<List<Integer>>results, List<Integer> subset, int[] nums, int startIdx) {
        results.add(new ArrayList<>(subset));
        for (int i = startIdx; i < nums.length; ++i) {
            subset.add(nums[i]);
            backtrace(results, subset, nums, i + 1);
            subset.remove(subset.size() - 1);
        }
    }
}